### PR TITLE
Roll up uncompressed chunks into compressed ones

### LINKED ISSUE #4079: [Enhancement]: Roll up multiple uncompressed chunks into single compressed chunk
### What type of enhancement is this?

Performance

### What subsystems and features will be improved?

Compression, Partitioning

### What does the enhancement do?

For high-ingest use cases (e.g., Promscale), it would be beneficial to be able to have small chunk sizes for uncompressed data in order to keep, e.g., index sizes small. However, having many small chunks can be bad for query performance so to offset that performance issue it would be good to be able to roll up several uncompressed chunks into one compressed chunk, which will still have a relatively small footprint in terms of disk space.

The Promscale team should be able to give more information on the performance benefits.

### Implementation challenges

Check with @cevian and @antekresic for recommendations on implementation.

### Why

When you combine the compressed chunks the performance should be better.

### How

- [ ] ...

### PR DESCRIPTION
**Pull request by antekresic**
This change introduces a new option to the compression procedure which decouples the uncompressed chunk interval from the compressed chunk interval. It does this by allowing multiple uncompressed chunks into one compressed chunk as part of the compression procedure. The main use-case is to allow much smaller uncompressed chunks than compressed ones. This has several advantages:
- Reduce the size of btrees on uncompressed data (thus allowing faster inserts because those indexes are memory-resident).
- Decrease disk-space usage for uncompressed data.
- Reduce number of chunks over historical data.

From a UX point of view, we simple add a compression with clause option `compress_chunk_time_interval`. The user should set that according to their needs for constraint exclusion over historical data. Ideally, it should be a multiple of the uncompressed chunk interval and so we throw a warning if it is not.

Fixes #4079

### CODE REVIEW

**Comment by cevian** [comment]
Patch has previously been reviewed and approved by me. Added @mkindahl to review since he was part of discussion. @mkindahl please add anyone else who you think is relevant to review.

File: sql/updates/latest-dev.sql [hunk 1]
```diff
@@ -46,3 +46,6 @@ BEGIN
     WHERE mat_hypertable_id OPERATOR(pg_catalog.=) _cagg_data.mat_hypertable_id;
 END;
 $BODY$;
+
+ALTER TABLE _timescaledb_catalog.dimension ADD COLUMN compress_interval_length bigint NULL;
+ALTER TABLE _timescaledb_catalog.dimension ADD CONSTRAINT dimension_compress_interval_length_check CHECK (compress_interval_length IS NULL OR compress_interval_length > 0);
```
Comment (svenklemm) [inline sql/updates/latest-dev.sql:51]: Doing catalog table changes like that is problematic and will lead to different schema in new installations vs upgraded installation. You need to rebuild any catalog table you intend to update to prevent that. There are plenty of examples in previous update scripts.

File: sql/updates/reverse-dev.sql [hunk 2]
```diff
@@ -45,3 +45,5 @@ BEGIN
     WHERE mat_hypertable_id OPERATOR(pg_catalog.=) _cagg_data.mat_hypertable_id;
 END;
 $BODY$;
+
+ALTER TABLE _timescaledb_catalog.dimension DROP COLUMN compress_interval_length;
```
Comment (svenklemm) [inline sql/updates/reverse-dev.sql:49]: Doing catalog table changes like that is problematic and will lead to different schema in new installations vs upgraded installation. You need to rebuild any catalog table you intend to update to prevent that. There are plenty of examples in previous update scripts.

File: src/chunk.c [hunk 3]
```diff
@@ -4617,6 +4617,72 @@ add_foreign_table_as_chunk(Oid relid, Hypertable *parent_ht)
...
 
+void
+ts_chunk_merge_across_dimension(Chunk *chunk, const Chunk *merge_chunk, int32 dimension_id)
+{
+	const DimensionSlice *slice, *merge_slice;
+	int num_ccs, i;
+	bool dimension_slice_found = false;
+
+	if (chunk->hypertable_relid != merge_chunk->hypertable_relid)
+		elog(ERROR, "cannot merge non-adjacent chunks from different hypertables");
```
Comment (svenklemm) [inline src/chunk.c:4628]:
```suggestion
		elog(ERROR, "cannot merge chunks from different hypertables");
```

File: tsl/test/sql/compression_merge.sql [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+\ir include/rand_generator.sql
+\c :TEST_DBNAME :ROLE_SUPERUSER
+\ir include/compression_utils.sql
+\c :TEST_DBNAME :ROLE_DEFAULT_PERM_USER
+
+CREATE TABLE test1 ("Time" timestamptz, i integer, value integer);
+SELECT table_name from create_hypertable('test1', 'Time', chunk_time_interval=> INTERVAL '1 hour');
+
+-- This will generate 24 chunks
+INSERT INTO test1 SELECT t, gen_rand_minstd(), gen_rand_minstd() FROM generate_series('2018-03-02 1:00'::TIMESTAMPTZ, '2018-03-03 0:59', '1 minute') t;
```
Comment (svenklemm) [inline tsl/test/sql/compression_merge.sql:14]: why do you need random data here?

File: src/dimension.c [hunk 5]
```diff
@@ -879,6 +897,23 @@ ts_dimension_set_chunk_interval(Dimension *dim, int64 chunk_interval)
...
+interval_to_usec(Interval *interval)
+{
+	return (interval->month * DAYS_PER_MONTH * USECS_PER_DAY) + (interval->day * USECS_PER_DAY) +
+		   interval->time;
+}
+
+int
+ts_dimension_set_compress_interval(Dimension *dim, int64 compress_interval)
+{
+	Assert(IS_OPEN_DIMENSION(dim));
```
Comment (mkindahl) [inline src/dimension.c:910]: Since this is reading from the metadata, it might make sense to generate an error if the dimension is not open. Otherwise, you might proceed with this code for release builds, causing hard-to-find bugs.

**Review by mkindahl — COMMENTED** [review]
Here is a first batch of comments. I will continue reviewing, but you might want to have a look at these first.

File: src/chunk.c [hunk 6]
```diff
@@ -4617,6 +4617,72 @@ add_foreign_table_as_chunk(Oid relid, Hypertable *parent_ht)
...
+		ts_chunk_constraint_scan_by_dimension_slice_id(slice->fd.id, NULL, CurrentMemoryContext);
+
+	Assert(num_ccs > 0);
+
+	DimensionSlice *new_slice =
+		ts_dimension_slice_create(dimension_id, slice->fd.range_start, merge_slice->fd.range_end);
+	if (num_ccs == 1)
+	{
+		ts_dimension_slice_delete_by_id(slice->fd.id, false);
+	}
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:4662]: Could you please add a comment to that effect?

File: sql/updates/tags [hunk 7]
```diff
@@ -0,0 +1 @@
+./../../tags
```
Comment (svenklemm) [inline sql/updates/tags:1]: don't think this belongs into this PR

File: tsl/src/compression/compression.c [hunk 8]
```diff
@@ -530,6 +532,50 @@ static SegmentInfo *segment_info_new(Form_pg_attribute column_attr);
 static void segment_info_update(SegmentInfo *segment_info, Datum val, bool is_null);
 static bool segment_info_datum_is_in_group(SegmentInfo *segment_info, Datum datum, bool is_null);
 
+static int32
+get_max_sequence_number(Relation compressed_table, int16 seq_num_column_num)
```
Comment (svenklemm) [inline tsl/src/compression/compression.c:536]: the sequence number is not global but local per segmentby group so any function assuming it is global is not handling sequence number correctly

File: .github/gh_matrix_builder.py [hunk 9]
```diff
@@ -66,6 +66,7 @@ def build_release_config(overrides):
     "build_type": "Release",
     "pg_build_args": "",
     "tsdb_build_args": "-DWARNINGS_AS_ERRORS=ON -DREQUIRE_ALL_TESTS=ON",
+    "installcheck_args": "IGNORES='bgw_db_scheduler bgw_db_scheduler_fixed chunk_merge'",
```
Comment (svenklemm) [inline .github/gh_matrix_builder.py:69]: Why does chunk_merge not work in a release configuration? I think this should not be part of the PR and investigated why it doesnt work.

File: tsl/src/compression/api.c [hunk 10]
```diff
@@ -199,18 +280,131 @@ restore_autovacuum_on_decompress(Oid uncompressed_hypertable_relid, Oid uncompre
 	}
 }
 
-static void
+Chunk *
+find_chunk_to_merge_into(Hypertable *ht, Chunk *current_chunk)
```
Comment (svenklemm) [inline tsl/src/compression/api.c:284]: should be static and not exported since you only call in this file

File: tsl/src/compression/api.c [hunk 11]
```diff
@@ -199,18 +280,131 @@ restore_autovacuum_on_decompress(Oid uncompressed_hypertable_relid, Oid uncompre
...
+	{
+		val = current_chunk->cube->slices[i]->fd.range_start;
+		if (current_chunk->cube->slices[i]->fd.dimension_id == time_dim->fd.id)
+		{
+			current_chunk_interval = current_chunk->cube->slices[i]->fd.range_end -
+									 current_chunk->cube->slices[i]->fd.range_start;
+			val--;
+		}
+		p->coordinates[p->num_coords++] = val;
+	}
```
Comment (svenklemm) [inline tsl/src/compression/api.c:312]:
```suggestion
  if (current_chunk->cube->slices[i]->fd.dimension_id != time_dim->fd.id)
  {
    elog(ERROR,"time dimension is not primary dimension");
  }
  p->coordinates[p->num_ccords++] = current_chunk->cube->slices[0]->fd.range_start - 1;
	for (int i = 1; i < current_chunk->cube->num_slices; i++)
	{
		p->coordinates[p->num_coords++] = current_chunk->cube->slices[i]->fd.range_start;
	}
```

The time dimension the merging happens on will always be the first dimension so i think this code could be simplified a bit, we can get time val and calculate new value and always copy the rest.

File: tsl/src/compression/api.c [hunk 12]
```diff
@@ -199,18 +280,131 @@ restore_autovacuum_on_decompress(Oid uncompressed_hypertable_relid, Oid uncompre
...
+		{
+			compressed_chunk_interval = previous_chunk->cube->slices[i]->fd.range_end -
+										previous_chunk->cube->slices[i]->fd.range_start;
+		}
+		else if (previous_chunk->cube->slices[i]->fd.id != current_chunk->cube->slices[i]->fd.id)
+		{
+			/* If the slices do not match (except on time dimension), we cannot merge the chunks. */
+			return NULL;
+		}
+	}
```
Comment (svenklemm) [inline tsl/src/compression/api.c:338]: This can be simplified similar to previous for-loop.

File: tsl/src/compression/api.c [hunk 13]
```diff
@@ -199,18 +280,131 @@ restore_autovacuum_on_decompress(Oid uncompressed_hypertable_relid, Oid uncompre
...
+
+/* Check if compression order is violated by merging in a new chunk
+ * Because data merged in uses higher sequence numbers than any data already in the chunk,
+ * the only way the order is guaranteed can be if we know the data we are merging in would come
+ * after the existing data according to the compression order. This is true if the data being merged
+ * in has timestamps greater than the existing data and the first column in the order by is time
+ * ASC.
+ */
+bool
+check_is_chunk_order_violated_by_merge(
```
Comment (svenklemm) [inline tsl/src/compression/api.c:356]: should be static function

File: src/chunk.c [hunk 14]
```diff
@@ -4617,6 +4617,107 @@ add_foreign_table_as_chunk(Oid relid, Hypertable *parent_ht)
...
+	bool dimension_slice_found = false;
+
+	if (chunk->hypertable_relid != merge_chunk->hypertable_relid)
+		ereport(ERROR,
+				(errmsg("cannot merge chunks from different hypertables"),
+				 errhint("chunk 1: \"%s\", chunk 2: \"%s\"",
+						 get_rel_name(chunk->table_id),
+						 get_rel_name(merge_chunk->table_id))));
+
+	for (int i = 0; i < chunk->cube->num_slices; i++)
```
Comment (svenklemm) [inline src/chunk.c:4634]: The code currently only support merging on primary time dimension which is always first dimension so the code should reflect that.

File: src/chunk.h [hunk 15]
```diff
@@ -231,6 +231,8 @@ extern void ts_chunk_scan_iterator_set_chunk_id(ScanIterator *it, int32 chunk_id
 extern bool ts_chunk_lock_if_exists(Oid chunk_oid, LOCKMODE chunk_lockmode);
 extern int ts_chunk_oid_cmp(const void *p1, const void *p2);
 int ts_chunk_get_osm_chunk_id(int hypertable_id);
+extern TSDLLEXPORT void ts_chunk_merge_across_dimension(Chunk *chunk, const Chunk *empty_chunk,
+														int32 dimension_id);
```
Comment (svenklemm) [inline src/chunk.h:235]: i think i would slightly prefer ts_chunk_merge_on_dimension

File: test/src/chunk.c [hunk 16]
```diff
@@ -0,0 +1,32 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
```
Comment (svenklemm) [inline test/src/chunk.c:2]: I dont see this being used anywhere, so if its dead code should be deleted.

File: tsl/src/compression/compression.c [hunk 17]
```diff
@@ -530,6 +555,205 @@ static SegmentInfo *segment_info_new(Form_pg_attribute column_attr);
...
+
+			num_segmentby_columns++;
+		}
+
+		/* Check that we have the correct number of index attributes
+		 * and that the last one is the sequence number
+		 */
+		if (num_segmentby_columns != index_rel->rd_att->natts - 1 ||
+			0 != namestrcmp((Name) &index_rel->rd_att->attrs[num_segmentby_columns].attname,
+							COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME))
```
Comment (svenklemm) [inline tsl/src/compression/compression.c:607]:
```suggestion
			namestrcmp((Name) &index_rel->rd_att->attrs[num_segmentby_columns].attname,
							COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME)  != 0)
```

File: tsl/src/compression/compression.c [hunk 18]
```diff
@@ -530,6 +555,205 @@ static SegmentInfo *segment_info_new(Form_pg_attribute column_attr);
...
+ * SEQUENCE_NUM_GAP.
+ */
+static int32
+get_sequence_number_for_current_group(Relation table_rel, Oid index_oid,
+									  int16 *uncompressed_col_to_compressed_col,
+									  PerColumn *per_column, int n_input_columns)
+{
+	/* If no index, relation must be empty so just reset. */
+	if (index_oid == InvalidOid)
+		return SEQUENCE_NUM_GAP;
```
Comment (svenklemm) [inline tsl/src/compression/compression.c:631]: That is not a correct assumption, the presence of an index is not dependent on there being data and vice versa.

File: tsl/src/compression/compression.c [hunk 19]
```diff
@@ -530,6 +555,205 @@ static SegmentInfo *segment_info_new(Form_pg_attribute column_attr);
...
+							   &is_null);
+		if (is_null)
+			result = 0;
+	}
+
+	index_endscan(index_scan);
+	index_close(index_rel, AccessShareLock);
+
+	MemoryContextSwitchTo(old_ctx);
+	MemoryContextReset(index_scan_ctx);
```
Comment (svenklemm) [inline tsl/src/compression/compression.c:704]:
```suggestion
	MemoryContextDelete(index_scan_ctx);
```

File: tsl/src/compression/compression.c [hunk 20]
```diff
@@ -631,21 +856,46 @@ row_compressor_init(RowCompressor *row_compressor, TupleDesc uncompressed_tuple_
...
+	row_compressor->index_oid =
+		get_compressed_chunk_index(compressed_table,
+								   row_compressor->uncompressed_col_to_compressed_col,
+								   row_compressor->per_column,
+								   row_compressor->n_input_columns);
+
+	/* We expect the appropriate index to exist for segment by compression
+	 * if the table is not empty. If it is empty, this is not a problem
+	 * since we are not merging chunks.
+	 */
```
Comment (svenklemm) [inline tsl/src/compression/compression.c:887]: Hmm why is index a requirement here, we can fall back to sequential scan if no index is present

File: test/sql/chunk_merge.sql [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+-- NOTE: we are merging the same chunk just so they have the exact same partitioning schema and we don't hit the previous test error.
+SELECT _timescaledb_internal.test_merge_chunks_across_dimension('_timescaledb_internal._hyper_1_1_chunk','_timescaledb_internal._hyper_1_1_chunk', 999);
+
+\set ON_ERROR_STOP 1
+
+-- Merge on open (time) dimension.
+SELECT _timescaledb_internal.test_merge_chunks_across_dimension('_timescaledb_internal._hyper_1_5_chunk','_timescaledb_internal._hyper_1_6_chunk', 1);
+
+-- Merge on close dimension.
+SELECT _timescaledb_internal.test_merge_chunks_across_dimension('_timescaledb_internal._hyper_1_1_chunk','_timescaledb_internal._hyper_1_4_chunk', 2);
```
Comment (svenklemm) [inline test/sql/chunk_merge.sql:52]: this dimension is not a closed dimension, i thought we only supported merging on primary dimension?

File: test/sql/chunk_merge.sql [hunk 22]
```diff
@@ -0,0 +1,54 @@
+-- This file and its contents are licensed under the Apache License 2.0.
```
Comment (svenklemm) [inline test/sql/chunk_merge.sql:1]: This file seems misplaced since compression is TLS-feature the test should  be in tsl folder and not mixed with apache-code

File: tsl/test/sql/chunk_merge.sql [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+-- Cannot merge chunks on with non-existant dimension slice.
+-- NOTE: we are merging the same chunk just so they have the exact same partitioning schema and we don't hit the previous test error.
+SELECT _timescaledb_internal.test_merge_chunks_on_dimension('_timescaledb_internal._hyper_1_1_chunk','_timescaledb_internal._hyper_1_1_chunk', 999);
+
+\set ON_ERROR_STOP 1
+
+-- Merge on open (time) dimension.
+SELECT _timescaledb_internal.test_merge_chunks_on_dimension('_timescaledb_internal._hyper_1_5_chunk','_timescaledb_internal._hyper_1_6_chunk', 1);
+
+-- Merge on closed dimension.
```
Comment (svenklemm) [inline tsl/test/sql/chunk_merge.sql:51]:
```suggestion
-- Merge on non-primary dimension.
```
This is not a closed dimension

File: tsl/test/sql/compression_merge.sql [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+\set SEGMENT_META_COL_MAX _ts_meta_max_1
+\ir include/compression_test_hypertable_segment_meta.sql
+
+DROP TABLE test2;
+
+CREATE TABLE test3 ("Time" timestamptz, i integer, loc integer, value integer);
+SELECT table_name from create_hypertable('test3', 'Time', chunk_time_interval=> INTERVAL '1 hour');
+SELECT table_name from  add_dimension('test3', 'i', chunk_time_interval=> 1);
+
+-- This will generate 25 1 hour chunks with a closed space dimension.
```
Comment (svenklemm) [inline tsl/test/sql/compression_merge.sql:62]: This is not a closed dimension.
```suggestion
-- This will generate 25 1 hour chunks.
```

Comment (antekresic) [reply tsl/test/sql/compression_merge.sql:62]: Changed the test to add a closed dimension.

File: tsl/src/compression/compression.c (hunk 20 above)
Comment (antekresic) [reply tsl/src/compression/compression.c]: Fixed so its not a requirement.

File: tsl/src/compression/api.c [hunk 25]
```diff
@@ -199,18 +280,138 @@ restore_autovacuum_on_decompress(Oid uncompressed_hypertable_relid, Oid uncompre
...
+		elog(ERROR, "hypertable has no open partitioning dimension");
+
+	if (time_dim->fd.compress_interval_length == 0)
+		return NULL;
+
+	if (current_chunk->cube->num_slices == 0 ||
+		current_chunk->cube->slices[0]->fd.dimension_id != time_dim->fd.id)
+	{
+		elog(ERROR, "primary dimension of the chunk is not the time dimension");
+	}
```
Comment (svenklemm) [inline tsl/src/compression/api.c:302]:
```suggestion
	Assert(current_chunk->cube->slices[0]->fd.dimension_id == time_dim->fd.id);
```
You cannot create a hypertable with with time not being primary dimension with our API

File: tsl/test/sql/chunk_merge.sql [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+-- Cannot merge same chunk to itself (its not adjacent to itself).
+SELECT _timescaledb_internal.test_merge_chunks_on_dimension('_timescaledb_internal._hyper_1_1_chunk','_timescaledb_internal._hyper_1_1_chunk', 1);
+
+-- Cannot merge chunks on with different partitioning schemas.
+SELECT _timescaledb_internal.test_merge_chunks_on_dimension('_timescaledb_internal._hyper_1_1_chunk','_timescaledb_internal._hyper_1_4_chunk', 1);
+
+-- Cannot merge chunks on with non-existant dimension slice.
+-- NOTE: we are merging the same chunk just so they have the exact same partitioning schema and we don't hit the previous test error.
+SELECT _timescaledb_internal.test_merge_chunks_on_dimension('_timescaledb_internal._hyper_1_1_chunk','_timescaledb_internal._hyper_1_1_chunk', 999);
+
```
Comment (svenklemm) [inline tsl/test/sql/chunk_merge.sql:45]: Please add a test where you try to merge chunks from the internal compressed hypertable (should error since it doesnt have dimensions)
Also please add a test for cagg hypertable (should work)

File: tsl/src/compression/api.c (hunk 25 above)
Comment (antekresic) [reply tsl/src/compression/api.c:302]: I wasn't able to find a way to create a hypertable without primary open dimension. Definitely not possible through the API.

File: tsl/test/sql/chunk_merge.sql (hunk 26 above)
Comment (antekresic) [reply tsl/test/sql/chunk_merge.sql:46]: Added.

**Review by svenklemm — APPROVED** [review]

File: tsl/src/compression/compression.c [hunk 27]
```diff
@@ -530,6 +555,226 @@ static SegmentInfo *segment_info_new(Form_pg_attribute column_attr);
...
+		/* Index scan should always use at least one scan key to get the sequence number. */
+		Assert(num_scankeys > 0);
+
+		result = index_scan_sequence_number(table_rel, index_oid, scankey, num_scankeys);
+	}
+	else
+	{
+		/* Table scan can work without scan keys. */
+		result = table_scan_sequence_number(table_rel, seq_num_column_num, scankey, num_scankeys);
+	}
```
Comment (mkindahl) [inline tsl/src/compression/compression.c:770]: There seems to be no test case for this situation.

File: src/chunk.c [hunk 28]
```diff
@@ -4665,6 +4665,116 @@ add_foreign_table_as_chunk(Oid relid, Hypertable *parent_ht)
...
+
+	num_ccs =
+		ts_chunk_constraint_scan_by_dimension_slice_id(slice->fd.id, NULL, CurrentMemoryContext);
+
+	if (num_ccs <= 0)
+		ereport(ERROR,
+				(errmsg("missing chunk constraint for dimension slice"),
+				 errhint("chunk: \"%s\", slice ID %d",
+						 get_rel_name(chunk->table_id),
+						 slice->fd.id)));
```
[the remark this answers is not in the document]
Comment (antekresic) [reply src/chunk.c:4736]: Definitely not a normal case, will add a comment like suggested.

File: tsl/src/compression/compression.c (hunk 27 above)
Comment (antekresic) [reply tsl/src/compression/compression.c:771]: There was one but the code wasn't scanning the data properly hence not getting coverage. I've fixed the code and the test to catch it.

File: src/chunk.c (hunk 28 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/chunk.c:4736]: Looks good.

We have added a macro `Ensure` that can be used for situations like this. It behaves as an `Assert` in debug builds so that you get a segfault and can investigate the issue, but print an error in release builds so that you can capture the case without a server crash in the wild.

**Review by mkindahl — APPROVED** [review]
LGTM. Added a comment below, but nothing you need to handle in this PR.
